tests/mock_crust/client_restrictions.rs

```diff
@@ -215,3 +219,132 @@ fn resend_over_load() {
     // Routing client will not send any notification regarding this expiration.
     assert!(clients[0].inner.try_next_ev().is_err());
 }
+
+// Sends the requests and verifies recipients if `rate_limiter` approves the request.
+// Returns rejected requests' message id and its client's ip address.
+fn rate_limiter_send_reqs(
+    network: &Network<PublicId>,
+    nodes: &mut [TestNode],
+    clients: &mut [TestClient],
+    total_usage: &mut u64,
+    is_always_send: bool,
+) -> HashMap<MessageId, IpAddr> {
+    let mut rng = network.new_rng();
+    let data_id: XorName = rng.gen();
+    let dst = Authority::NaeManager(data_id);
+
+    let mut clients_sent = HashMap::new();
+    for client in clients.iter_mut() {
+        if is_always_send || rng.gen_weighted_bool(2) {
+            let msg_id = MessageId::new();
+            unwrap!(client.inner.get_idata(dst, data_id, msg_id));
+            let _ = clients_sent.insert(msg_id, client.ip());
+        }
+    }
+    trace!("clients_sent: {:?}", clients_sent);
+    let _ = poll_all(nodes, clients);
+
+    let mut request_received: HashMap<MessageId, usize> = HashMap::new();
+    for node in nodes.iter_mut().filter(|n| n.is_recipient(&dst)) {
+        while let Ok(event) = node.try_next_ev() {
+            if let Event::Request {
+                request: Request::GetIData { msg_id: req_message_id, .. }, ..
+            } = event
+            {
+                let entry = request_received.entry(req_message_id).or_insert(0);
+                *entry += 1;
+            }
+        }
+    }
+    trace!("request_received: {:?}", request_received);
+
+    for (msg_id, count) in &request_received {
+        assert_eq!(*count, MIN_SECTION_SIZE);
+        let _ = clients_sent.remove(msg_id);
+        *total_usage += MAX_IMMUTABLE_DATA_SIZE_IN_BYTES;
+    }
+
+    // Max bandwidth used is C + C/2 + C/3 + C/4 + ... + C/2 = C * HarmonicNumber(N)
+    // FIXME(michael): this isn't really the max bandwidth, it can get higher somehow,
+    // but I haven't worked out how yet.
+    let cap = ::std::cmp::max(
+        SOFT_CAPACITY * harmonic_number(clients.len()).ceil() as u64,
+        MIN_CLIENT_CAPACITY * clients.len() as u64,
+    );
+    assert!(*total_usage <= cap, "{} > {}", total_usage, cap);
```

Comment:
It seems to fail on about the 8th iteration each time, which makes me think that the bucket fills up with junk as more iterations run, and because there's no hard cap, the usage keeps growing (potentially) indefinitely.